lib/chef_zero/server.rb

```diff
@@ -238,67 +240,78 @@ def start(publish = true)
     # to the forked process.
     #
     # @param [Fixnum] wait
-    #   the number of seconds to wait for the server to start
+    #   the number of seconds to wait for the server to start (nil or 0 means forever)
     #
     # @return [Thread]
     #   the thread the background process is running in
     #
-    def start_background(wait = 5)
-      @server = WEBrick::HTTPServer.new(
-        :DoNotListen => true,
-        :AccessLog   => [],
-        :Logger      => WEBrick::Log.new(StringIO.new, 7),
-        :SSLEnable  => options[:ssl],
-        :SSLCertName  => [ [ 'CN', WEBrick::Utils::getservername ] ],
-        :StartCallback => proc {
-          @running = true
-        }
-      )
+    # @raise [ServerAlreadyStartedError] if the server is already started.
+    #
+    def start_background(wait = nil)
+      server_mutex.synchronize do
+        if !server.nil?
+          raise ServerAlreadyStartedError
+        end
+        self.server = WEBrick::HTTPServer.new(
+          :DoNotListen => true,
+          :AccessLog   => [],
+          :Logger      => WEBrick::Log.new(StringIO.new, 7),
+          :SSLEnable  => options[:ssl],
+          :SSLCertName  => [ [ 'CN', WEBrick::Utils::getservername ] ],
+          :StartCallback => proc {
+            self.running = true
+          }
+        )
+      end
+
       ENV['HTTPS'] = 'on' if options[:ssl]
-      @server.mount('/', Rack::Handler::WEBrick, app)
+      server.mount('/', Rack::Handler::WEBrick, app)
 
       # Pick a port
       if options[:port].respond_to?(:each)
+        got_port = nil
         options[:port].each do |port|
           begin
-            @server.listen(options[:host], port)
-            @port = port
+            server.listen(options[:host], port)
+            self.port = got_port = port
             break
           rescue Errno::EADDRINUSE
             ChefZero::Log.info("Port #{port} in use: #{$!}")
           end
         end
-        if !@port
+        if !got_port
           raise Errno::EADDRINUSE, "No port in :port range #{options[:port]} is available"
         end
       else
-        @server.listen(options[:host], options[:port])
-        @port = options[:port]
+        server.listen(options[:host], options[:port])
+        self.port = options[:port]
       end
 
       # Start the server in the background
-      @thread = Thread.new do
+      self.thread = Thread.new do
         begin
           Thread.current.abort_on_exception = true
-          @server.start
+          server.start
         ensure
-          @port = nil
-          @running = false
+          self.port = nil
+          self.running = false
         end
       end
 
       # Do not return until the web server is genuinely started.
-      while !@running && @thread.alive?
-        sleep(0.01)
+      Timeout.timeout(wait) do
```

Comment:
Is raising a `Timeout::Error` okay, or should we wrap this in a custom exception? Not even a suggestion, but just a question :smile: